Feat: post 본문 무한 스크롤 구현

ckswnskfk opened the pull request:
## 어떤 변경인지 
- [x] Feat: 기능 변경, 기능 추가
- [ ] Docs: 문서 작업
- [ ] Refactor: 기능 변경 없이 코드 수정
- [ ] Style: 디자인
- [ ] Fix: 버그 수정
- [ ] Test: 테스트 코드 작성

## 설명
인터넷 여기 저기 참고해서 post 페이지 무한 스크롤을 구현했습니다.
그리고 해상도에 따라 데이터 fetching 갯수(limit) 값을 다르게 하는 아이디어도 있으니 한 번 봐주세요!

### 이슈 번호
https://github.com/Codeit-Rolling-11-Letsgo/Rolling/issues/37

### 주요 변경 사항 
1. IntersectionObserver 를 사용할 수 있는 custom hook 생성
2. 해상도에 따라 fetching 갯수(limit) 값 리턴하는 함수 생성

### 어떻게 동작하는지

### To Reviewers

ckswnskfk committed:
Refactor: ky의 searchParams 옵션 사용으로 코드량 줄이고 가독성 올림

ckswnskfk committed:
Feat: post 페이지 무한 스크롤 구현

ckswnskfk committed:
Refactor: 변수명 변경 및 jsdoc 추가

ckswnskfk committed:
Style: padding-bottom 조정

ckswnskfk committed:
Refactor: 유틸 함수 파일명 변경

ckswnskfk committed:
Feat: 해상도에 따라 초기 limit 값을 리턴하는 함수 생성

ckswnskfk commented on src/apis/post/postAPI.js:
```diff
@@ -1,14 +1,24 @@
-
-	const response = await fetcher
-		.get(`recipients/${recipientId}/messages/?${params}`)
+	const messagesInfo = await fetcher
+		.get(`recipients/${recipientId}/messages/`, {
+			searchParams: { limit, offset },
+		})
 		.json();
```
용준님께서 말씀하신 ky의 옵션이 이거였죠.. 편합니다요 👍

mizign replied:
퍼갑니다요 🤗

ckswnskfk commented on src/utils/commonUtils.js:
아이디어는 괜찮은 거 같은데 코드가 좀 부실해보이네요..;
의견 부탁드립니다!

mizign commented on src/utils/commonUtils.js:
```diff
@@ -0,0 +1,7 @@
+export const getLimitByResolution = () => {
+	const width = window.innerWidth;
+
+	if (width > 1248) return 8;
+	if (width > 767) return 5;
+	return 3;
+};
```
이 아이디어 너무 획기적이라고 생각합니다 ㅋㅋ 그런데 어쨌든 이것도 뷰포트에 따른 반응형 구현이니까, 미디어 쿼리를 사용하는게 유지보수 측면에서 좋지 않을까하고 제안드려봅니다. (media라는 키워드로 검색이 되도록요) react-responsive 라이브러리의 useMediaQuery 라는 훅을 사용해서 자바스크립트 내에서 뷰포트에 따른 반응형 코드를 짠 적이 있는데 이것도 참고하시면 좋을 것 같아요! 물론 짧은코드라 새 라이브러리를 사용하는 것이 득보다 실이 많을수도 있겠네요,,

ckswnskfk replied:
아직 break point도 정해지지 않아 생각 못했는데 좋네요! 👍 
저도 천천히 알아보겠습니다ㅎㅎ

mizign commented on src/pages/post/PostPage.jsx:
```diff
@@ -1,21 +1,55 @@
+		}
+	};
+
+	const intersectionOptions = {
+		rootMargin: '350px',
+		threshold: 0,
+	};
```
그런데 useIntersect 훅에 옵션을 아예 안 넣어도 잘 동작하는 것 같아요. threshold는 디폴트가 0인것 같고, 로딩상태에 따라 트리거를 보이고 안보이고를 설정해주셨으니까 루트 마진 속성도 필요 없을 것 같은데 혹시 옵션의 의도가 있으셨을까요? 👀

ckswnskfk replied:
아 threshold는 필요 없겠네요!
rootMargin의 경우 사용자가 가장 아래까지 스크롤 하기 전에 로딩이 이뤄지면 좋겠어서 대충 카드 하나 크기만큼 지정한거랍니다ㅋㅋ

mizign replied:
아 그렇겠네요 옵저버도 div로 배치되니까..! 👍 세심하시군요

mizign reviewed (commented):
무한스크롤 이름만 들어봤지 실제로 구현된 코드는 처음 보는데 너무 깔끔하게 써주셔서 공부하는 입장에서 보기가 편헀던 것 같습니다! 이 기회에 많이 배워가네요 👍
제가 아직 옵션 부분의 의도를 파악하지 못해서 그 부분에 대한 질문은 위에 리뷰로 남겼습니다. 나머지는 더 드릴 말씀이 없네요,, 수고하셨어요!

ckswnskfk committed:
Refactor: 리뷰 사항 반영

필요 없는 intersectionOptions 속성 삭제
getLimitByResolution 함수 내 matchMedia 함수 사용으로 미디어 쿼리 적용 및 media로 검색 가능하게 함

ckswnskfk committed:
Fix: PostPage -> PostIdPage 로 변경

dkile commented:
해상도에 따라 데이터 fetching 개수를 다르게 하면 어떤 장점이 있나요?

dkile commented on src/apis/post/postAPI.js:
```diff
@@ -1,14 +1,24 @@
+ *  }
+ * }>}
+ */
 export const getMessages = async ({ recipientId, limit, offset }) => {
```
저는 recipientList로 했는데 List랑 복수 s 중에서 하나로 통일을 하면 좋을 것 같네요. 어떤게 편하신가요?

dkile commented:
추후 util 함수들은 파일 네이밍을 고치긴 해야겠네요. mvp 구현 끝나고 한 번 잡아보시죠. 나머지는 모두 괜찮은 것 같습니다.

ckswnskfk committed:
Refactor: 리뷰 사항 반영

해상도에 따른 limit 반환하는 함수 삭제
배열 데이터 어미 List 붙이는 걸로 통일

ckswnskfk committed:
Refactor: 사용하지 않는 파일 삭제
